File: freeq-app/src/lib/act-facts.test.ts
import { describe, it, expect } from 'vitest';
import { actFacts, unknownFields } from './act-facts';

const R = (k: string) => (k === 'did:key:zWORKER' ? 'cardworker2' : k);

describe('the card facts', () => {
  it('a directed offer names its recipient, resolved, never raw', () => {
    expect(actFacts({ 'act-to': 'did:key:zWORKER' }, true, R)).toEqual([['offered to', 'cardworker2']]);
  });
  it('an opener with no recipient is offered to anyone', () => {
    expect(actFacts({}, true, R)).toEqual([['offered to', 'anyone']]);
  });
  it('a follow-up with no act-to claims nothing about audience', () => {
    expect(actFacts({}, false, R)).toEqual([]);
  });
  it('money is labelled: price on offers, bid on bids', () => {
    expect(actFacts({ 'act-price': '250 USD' }, true, R))
      .toEqual([['offered to', 'anyone'], ['price', '250 USD']]);
    expect(actFacts({ 'act-bid': '200 USD' }, false, R)).toEqual([['bid', '200 USD']]);
  });
  it('deadlines carry a local time as their value', () => {
    const facts = actFacts({ 'act-deadline': '1788000000' }, true, R);
    expect(facts[0]).toEqual(['offered to', 'anyone']);
    expect(facts[1][0]).toBe('deadline');
    expect(facts[1][1]).toBeTruthy();
  });
  it('a bid deadline gets its own label', () => {
    const facts = actFacts({ 'act-bid-deadline': '1788000000' }, true, R);
    expect(facts[1][0]).toBe('bids close');
  });
  it('capabilities are labelled as required skills', () => {
    expect(actFacts({ 'act-caps': 'url_fetch' }, true, R))
      .toEqual([['offered to', 'anyone'], ['skills required', 'url_fetch']]);
  });
  it("the award's winner gets an awarded-to line of its own", () => {
    expect(actFacts({}, false, R, 'did:key:zWORKER')).toEqual([['awarded to', 'cardworker2']]);
  });
  it('an unreadable deadline is skipped, not rendered as garbage', () => {
    expect(actFacts({ 'act-deadline': 'soon' }, false, R)).toEqual([]);
  });
  it('the note is a row of the grid, not a line under it', () => {
    expect(actFacts({ 'act-note': 'two days' }, false, R)).toEqual([['note', 'two days']]);
  });
  it('the context link is a row', () => {
    expect(actFacts({ 'act-ctx': 'https://example.org/a' }, false, R))
      .toEqual([['context', 'https://example.org/a']]);
  });
  it('the context hash is a row', () => {
    expect(actFacts({ 'act-ctx-h': 'sha256:9f00' }, false, R))
      .toEqual([['hash', 'sha256:9f00']]);
  });
  it('the payee is a row: a DID resolves, anything else is shown as sent', () => {
    expect(actFacts({ 'act-pay-to': 'did:key:zWORKER' }, false, R))
      .toEqual([['pay to', 'cardworker2']]);
    expect(actFacts({ 'act-pay-to': '0xdeadbeef' }, false, R))
      .toEqual([['pay to', '0xdeadbeef']]);
  });
  it('the payment is a row', () => {
    expect(actFacts({ 'act-tx': 'eth:0xdemo' }, false, R)).toEqual([['payment', 'eth:0xdemo']]);
  });
  it('the action a revision replaces is a row, under its raw id for now', () => {
    expect(actFacts({ 'act-replaces': '01JOLD' }, false, R)).toEqual([['replaces', '01JOLD']]);
  });
  it('the scope is a row', () => {
    expect(actFacts({ 'act-scope': 'room' }, false, R)).toEqual([['scope', 'room']]);
  });
  it('the seven follow the labelled facts, in their own fixed order', () => {
    expect(actFacts({
      'act-price': '250 USD', 'act-caps': 'url_fetch', 'act-note': 'two days',
      'act-ctx': 'https://example.org/a', 'act-ctx-h': 'sha256:9f00',
      'act-pay-to': 'did:key:zW', 'act-tx': 'eth:0xdemo', 'act-replaces': '01JOLD',
      'act-scope': 'room',
    }, true, R)).toEqual([
      ['offered to', 'anyone'],
      ['price', '250 USD'],
      ['skills required', 'url_fetch'],
      ['note', 'two days'],
      ['context', 'https://example.org/a'],
      ['hash', 'sha256:9f00'],
      ['pay to', 'did:key:zW'],
      ['payment', 'eth:0xdemo'],
      ['replaces', '01JOLD'],
      ['scope', 'room'],
    ]);
  });
});

describe('the unlabelled fields', () => {
  it('a field the card has no label for keeps its own key', () => {
    expect(unknownFields({ 'act-mystery': 'y', 'act-oracle': 'z' }))
      .toEqual([['mystery', 'y'], ['oracle', 'z']]);
  });
  it('the four newly labelled fields no longer fall through', () => {
    expect(unknownFields({
      'act-pay-to': 'did:key:zW', 'act-tx': 'eth:0xabc',
      'act-replaces': '01JOLD', 'act-scope': 'room',
    })).toEqual([]);
  });
  it('labelled and structural fields never fall through', () => {
    expect(unknownFields({
      act: 'handoff', 'act-verb': 'offer', 'act-id': 'X', 'act-to': 'd',
      'act-title': 't', 'act-note': 'n', 'act-ctx': 'u', 'act-ctx-h': 'h', 'act-deadline': '1',
      'act-bid-deadline': '1', 'act-caps': 'c', 'act-price': 'p',
      'act-bid': 'b', 'act-accepts': 'e', 'act-subject': 's',
      'act-pay-to': 'p2', 'act-tx': 'tx', 'act-replaces': 'r', 'act-scope': 'sc',
    })).toEqual([]);
  });
  it('non-act tags are not its business', () => {
    expect(unknownFields({ msgid: 'X', 'act-mystery': 'y' })).toEqual([['mystery', 'y']]);
  });
});

File: freeq-android/freeq/src/test/java/com/freeq/model/ActFactsTest.kt
package com.freeq.model

import org.junit.Assert.assertEquals
import org.junit.Test

class ActFactsTest {
    private val resolve = { k: String -> if (k == "did:key:zWORKER") "cardworker2" else k }

    @Test fun a_directed_offer_names_its_recipient_resolved() {
        assertEquals(
            listOf("offered to" to "cardworker2"),
            ActFacts.facts(mapOf("act-to" to "did:key:zWORKER"), true, resolve),
        )
    }

    @Test fun an_opener_is_offered_to_anyone_and_a_follow_up_claims_nothing() {
        assertEquals(listOf("offered to" to "anyone"), ActFacts.facts(emptyMap(), true, resolve))
        assertEquals(emptyList<Pair<String, String>>(), ActFacts.facts(emptyMap(), false, resolve))
    }

    @Test fun money_is_labelled_price_on_offers_and_bid_on_bids() {
        assertEquals(
            listOf("offered to" to "anyone", "price" to "250 USD"),
            ActFacts.facts(mapOf("act-price" to "250 USD"), true, resolve),
        )
        assertEquals(
            listOf("bid" to "200 USD"),
            ActFacts.facts(mapOf("act-bid" to "200 USD"), false, resolve),
        )
    }

    @Test fun deadlines_carry_a_time_value_and_garbage_is_skipped() {
        val f = ActFacts.facts(mapOf("act-deadline" to "1788000000"), true, resolve)
        assertEquals("offered to" to "anyone", f[0])
        assertEquals("deadline", f[1].first)
        assert(f[1].second.isNotEmpty())
        assertEquals(
            emptyList<Pair<String, String>>(),
            ActFacts.facts(mapOf("act-deadline" to "soon"), false, resolve),
        )
    }

    @Test fun a_bid_deadline_gets_its_own_label() {
        val f = ActFacts.facts(mapOf("act-bid-deadline" to "1788000000"), true, resolve)
        assertEquals("bids close", f[1].first)
    }

    @Test fun capabilities_are_labelled_as_required_skills() {
        assertEquals(
            listOf("offered to" to "anyone", "skills required" to "demo"),
            ActFacts.facts(mapOf("act-caps" to "demo"), true, resolve),
        )
    }

    @Test fun the_awards_winner_gets_an_awarded_to_line() {
        assertEquals(
            listOf("awarded to" to "cardworker2"),
            ActFacts.facts(emptyMap(), false, resolve, "did:key:zWORKER"),
        )
    }

    @Test fun the_note_is_a_row_of_the_grid_not_a_line_under_it() {
        assertEquals(
            listOf("note" to "two days"),
            ActFacts.facts(mapOf("act-note" to "two days"), false, resolve),
        )
    }

    @Test fun the_context_link_is_a_row() {
        assertEquals(
            listOf("context" to "https://example.org/a"),
            ActFacts.facts(mapOf("act-ctx" to "https://example.org/a"), false, resolve),
        )
    }

    @Test fun the_context_hash_is_a_row() {
        assertEquals(
            listOf("hash" to "sha256:9f00"),
            ActFacts.facts(mapOf("act-ctx-h" to "sha256:9f00"), false, resolve),
        )
    }

    @Test fun the_payee_is_a_row_a_did_resolves_and_anything_else_is_shown_as_sent() {
        assertEquals(
            listOf("pay to" to "cardworker2"),
            ActFacts.facts(mapOf("act-pay-to" to "did:key:zWORKER"), false, resolve),
        )
        assertEquals(
            listOf("pay to" to "0xdeadbeef"),
            ActFacts.facts(mapOf("act-pay-to" to "0xdeadbeef"), false, resolve),
        )
    }

    @Test fun the_payment_is_a_row() {
        assertEquals(
            listOf("payment" to "eth:0xdemo"),
            ActFacts.facts(mapOf("act-tx" to "eth:0xdemo"), false, resolve),
        )
    }

    @Test fun the_action_a_revision_replaces_is_a_row_under_its_raw_id() {
        assertEquals(
            listOf("replaces" to "01JOLD"),
            ActFacts.facts(mapOf("act-replaces" to "01JOLD"), false, resolve),
        )
    }

    @Test fun the_scope_is_a_row() {
        assertEquals(
            listOf("scope" to "room"),
            ActFacts.facts(mapOf("act-scope" to "room"), false, resolve),
        )
    }

    @Test fun the_seven_follow_the_labelled_facts_in_their_own_fixed_order() {
        assertEquals(
            listOf(
                "offered to" to "anyone",
                "price" to "250 USD",
                "skills required" to "url_fetch",
                "note" to "two days",
                "context" to "https://example.org/a",
                "hash" to "sha256:9f00",
                "pay to" to "did:key:zW",
                "payment" to "eth:0xdemo",
                "replaces" to "01JOLD",
                "scope" to "room",
            ),
            ActFacts.facts(
                mapOf(
                    "act-price" to "250 USD", "act-caps" to "url_fetch", "act-note" to "two days",
                    "act-ctx" to "https://example.org/a", "act-ctx-h" to "sha256:9f00",
                    "act-pay-to" to "did:key:zW", "act-tx" to "eth:0xdemo",
                    "act-replaces" to "01JOLD", "act-scope" to "room",
                ),
                true,
                resolve,
            ),
        )
    }

    @Test fun unlabelled_fields_keep_their_keys_and_known_ones_never_do() {
        assertEquals(listOf("mystery" to "y"), ActFacts.unknownFields(mapOf("act-mystery" to "y")))
        assertEquals(emptyList<Pair<String, String>>(), ActFacts.unknownFields(mapOf(
            "act-pay-to" to "did:key:zW", "act-tx" to "eth:0xabc",
            "act-replaces" to "01JOLD", "act-scope" to "room",
        )))
        assertEquals(emptyList<Pair<String, String>>(), ActFacts.unknownFields(mapOf(
            "act" to "handoff", "act-verb" to "offer", "act-id" to "X", "act-to" to "d",
            "act-title" to "t", "act-note" to "n", "act-ctx" to "u", "act-ctx-h" to "h", "act-deadline" to "1",
            "act-bid-deadline" to "1", "act-caps" to "c", "act-price" to "p",
            "act-bid" to "b", "act-accepts" to "e", "act-subject" to "s",
            "act-pay-to" to "p2", "act-tx" to "tx", "act-replaces" to "r", "act-scope" to "sc",
        )))
    }
}
